app/agent/runtime/qemu/qemu.service.ts

```diff
@@ -0,0 +1,542 @@
+    private readonly networkService: WorkspaceNetworkService,
+    public readonly instanceId: string,
+  ) {
+    this.agentConfig = config.agent();
+    this.instanceDir = `/srv/jailer/qemu/${instanceId}`;
+    this.instanceDirRoot = `${this.instanceDir}/root`;
+    this.vmInfoFilePath = `${this.instanceDirRoot}/vm-info.json`;
+    this.qmpSocketPath = path.join(this.instanceDirRoot, CHROOT_PATH_TO_SOCK);
+    if (this.qmpSocketPath.length > MAX_UNIX_SOCKET_PATH_LENGTH) {
+      // https://blog.8-p.info/en/2020/06/11/unix-domain-socket-length/
+      throw new Error(
+        "Instance ID is too long. Length of the path to the socket exceeds the maximum on Linux.",
+      );
+    }
+  }
+
+  private async getVirtioDeviceId(path: string) {
+    // https://github.com/cloud-hypervisor/cloud-hypervisor/blob/6acceaa87cfd6327d99d212e3854cabf72c608cc/block_util/src/lib.rs#L82
+    const st = await fs.stat(path);
+    return `${st.dev}${st.rdev}${st.ino}`.substring(0, VIRTIO_BLK_ID_BYTES);
+  }
+
+  private getVMLogsPath(): string {
+    return `/srv/jailer/qemu/logs/${this.instanceId}.log`;
+  }
+
+  private getStdinPath(): string {
+    return `/tmp/${this.instanceId}.stdin`;
+  }
+
+  private async writeVmInfoFile(args: { pid: number; ipBlockId: number }): Promise<void> {
+    const vmInfo: VmInfo = {
+      ...args,
+      instanceId: this.instanceId,
+    };
+    await fs.writeFile(this.vmInfoFilePath, JSON.stringify(vmInfo));
+  }
+
+  private async readVmInfoFile(): Promise<VmInfo> {
+    const contents = await fs.readFile(this.vmInfoFilePath);
+    return VmInfoValidator.Parse(JSON.parse(contents.toString()));
+  }
+
+  /**
+   * If the instance files are on disk, returns the result object.
+   * If the instance does not exist on disk, returns null.
+   * */
+  async getRuntimeInfo(): Promise<{ status: "on" | "off" | "paused"; info: VmInfo } | null> {
+    const info = await this.readVmInfoFile().catch((err) => {
+      if (err?.code === "ENOENT") {
+        return null;
+      }
+      throw err;
+    });
+    if (info === null) {
+      return null;
+    }
+
+    let sock: Socket;
+    try {
+      sock = await this.qmpConnect();
+    } catch (err: any) {
+      if (err instanceof QmpConnectionFailedError) return null;
+      throw err;
+    }
+
+    await this.sendQMPMessage(sock, "query-status");
+    const m = await this.waitForQMPMessageWithTimeout(sock);
+    if (m === void 0) {
+      return null;
+    }
+    sock.destroy();
+    return {
+      status: match((m as any).return.status)
+        .with("running", () => "on" as const)
+        .with("suspended", () => "paused" as const)
+        .with("paused", () => "paused" as const)
+        .with("postmigrate", () => "paused" as const)
+        .with("shutdown", () => "off" as const)
+        .with("internal-error", () => "off" as const)
+        .with("io-error", () => "off" as const)
+        .with(null, () => "off" as const)
+        .exhaustive(),
```

Comment:
This isn't actually exhaustive because m.return.status is of type any. Instead of `exhaustive` use `otherwise` and throw an error that such a status is unknown.